core/src/ledger/storage/masp_conversions.rs

```diff
@@ -90,15 +104,80 @@ where
+            if *addr == address::nam() {
+                // The amount that will be given of the new native token for
+                // every amount of the native token given in the
+                // previous epoch
+                let new_normed_inflation = *normed_inflation
+                    + (*normed_inflation * reward.0) / reward.1;
+                // The conversion is computed such that if consecutive
+                // conversions are added together, the
+                // intermediate native tokens cancel/
+                // telescope out
+                current_convs.insert(
+                    (addr.clone(), denom),
+                    (MaspAmount::from_pair(
+                        old_asset,
+                        -(*normed_inflation as i32),
+                    )
+                    .unwrap()
+                        + MaspAmount::from_pair(
+                            new_asset,
+                            new_normed_inflation as i32,
+                        )
+                        .unwrap())
+                    .into(),
+                );
+                // Operations that happen exactly once for each token
+                if denom == MaspDenom::Three {
+                    // The reward for each reward.1 units of the current asset
+                    // is reward.0 units of the reward token
+                    total_reward += (addr_bal
+                        * (new_normed_inflation, *normed_inflation))
+                        .0
+                        - addr_bal;
+                    // Save the new normed inflation
+                    *normed_inflation = new_normed_inflation;
+                }
+            } else {
+                // Express the inflation reward in real terms, that is, with
+                // respect to the native asset in the zeroth
+                // epoch
+                let real_reward =
+                    (reward.0 * ref_inflation) / *normed_inflation;
+                // The conversion is computed such that if consecutive
+                // conversions are added together, the
+                // intermediate tokens cancel/ telescope out
+                current_convs.insert(
+                    (addr.clone(), denom),
+                    (MaspAmount::from_pair(old_asset, -(reward.1 as i32))
                         .unwrap()
-                    + MaspAmount::from_pair(reward_asset, reward.0 as i32)
+                        + MaspAmount::from_pair(new_asset, reward.1 as i32)
+                            .unwrap()
+                        + MaspAmount::from_pair(
+                            reward_assets[denom as usize],
+                            real_reward as i32,
+                        )
                         .unwrap())
-                .into(),
-            );
+                    .into(),
+                );
+                // Operations that happen exactly once for each token
+                if denom == MaspDenom::Three {
```

Comment:
I don't think that the body of this conditional was called 4 times before. This `denom == MaspDenom::Three` check became necessary to stop repetitions once I had moved the logic occurring before the `for denom in token::MaspDenom::iter() {` loop into the `for denom in token::MaspDenom::iter() {` loop. I made this movement as a matter of convenience, but I'm sure that there are other ways this logic could be structured.

I chose `denom::Three` because the conditional's body modifies variables (namely `total_reward` and `normed_inflation`) that would affect the next iteration of `for denom in token::MaspDenom::iter() {`. In the current context, the `normed_inflation` can only be modified once we have finished working with the native token.